File: Fastmate/SettingsViewController.m
#import "SettingsViewController.h"
#import "UserDefaultsKeys.h"

@interface SettingsViewController ()

@property (nonatomic, weak) IBOutlet NSButton *defaultWatchedFoldersButton;
@property (nonatomic, weak) IBOutlet NSButton *allWatchedFoldersButton;
@property (nonatomic, weak) IBOutlet NSButton *specificWatchedFoldersButton;

@end

@implementation SettingsViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    WatchedFolderType type = [NSUserDefaults.standardUserDefaults integerForKey:WatchedFolderTypeKey];
    NSArray *buttons = @[self.defaultWatchedFoldersButton, self.allWatchedFoldersButton, self.specificWatchedFoldersButton];
    [buttons[type] setState:NSControlStateValueOn];
}

- (IBAction)watchedFolderButtonSelected:(NSButton *)sender {
    [NSUserDefaults.standardUserDefaults setInteger:sender.tag forKey:WatchedFolderTypeKey];
}

- (IBAction)openUserScriptsFolder:(id)sender {
    [NSWorkspace.sharedWorkspace openFile:[NSHomeDirectory() stringByAppendingPathComponent:@"userscripts"]];
}

- (IBAction)openUserStylesFolder:(id)sender {
    [NSWorkspace.sharedWorkspace openFile:[NSHomeDirectory() stringByAppendingPathComponent:@"userstyles"]];
}


@end

// This simply transforms the tag of a button to a bool, to check if the
// selected radio button is "Specific folders" in cocoa bindings.
@interface WatchedFolderTypeIsSpecificFolders: NSValueTransformer {} @end

@implementation WatchedFolderTypeIsSpecificFolders

+ (Class)transformedValueClass {
    return [NSNumber class];
}

+ (BOOL)allowsReverseTransformation {
    return NO;
}

- (id)transformedValue:(NSNumber *)value {
    return (value.integerValue == 2) ? @YES : @NO;
}

@end

File: Fastmate/SettingsViewController.swift
import Foundation
import Cocoa
import Combine

class SettingsViewController: NSViewController {

    @IBOutlet weak var watchedFolderTypeButtonInbox: NSButton!
    @IBOutlet weak var watchedFolderTypeButtonAll: NSButton!
    @IBOutlet weak var watchedFolderTypeButtonSpecific: NSButton!
    @IBOutlet weak var showUnreadInStatusBarButton: NSButton!
    @IBOutlet weak var showUnreadCountInDockButton: NSButton!
    @IBOutlet weak var userScriptsFolderButton: NSButton!
    @IBOutlet weak var watchedFoldersTextField: NSTextField!

    var subscriptions = Set<AnyCancellable>()

    override func viewDidLoad() {
        super.viewDidLoad()
        let settings = Settings.shared

        let watchedFolderType = settings.$watchedFolderType
            .map { WatchedFolderType(rawValue: $0) ?? .selected }
            .receive(on: DispatchQueue.main)

        watchedFolderType
            .sink { [weak self] in self?.set(watchedFolderType: $0) }
            .store(in: &subscriptions)

        watchedFolderType
            .map { $0 == .specific }
            .assign(to: \.isEnabled, on: watchedFoldersTextField)
            .store(in: &subscriptions)

        settings.$shouldShowStatusBarIcon
            .receive(on: DispatchQueue.main)
            .assign(to: \.isEnabled, on: showUnreadInStatusBarButton)
            .store(in: &subscriptions)

        settings.$shouldShowUnreadMailInDock
            .receive(on: DispatchQueue.main)
            .assign(to: \.isEnabled, on: showUnreadCountInDockButton)
            .store(in: &subscriptions)

        Publishers.Merge3(
            watchedFolderTypeButtonInbox.actionPublisher.map { WatchedFolderType.selected },
            watchedFolderTypeButtonSpecific.actionPublisher.map { WatchedFolderType.specific },
            watchedFolderTypeButtonAll.actionPublisher.map { WatchedFolderType.all })
            .map(\.rawValue)
            .assign(to: \.watchedFolderType, on: settings)
            .store(in: &subscriptions)

        userScriptsFolderButton.actionPublisher
            .map { (NSHomeDirectory() as NSString).appendingPathComponent("userscripts") }
            .sink { NSWorkspace.shared.openFile($0) }
            .store(in: &subscriptions)
    }

    func set(watchedFolderType: WatchedFolderType) {
        watchedFolderTypeButtonInbox?.state = watchedFolderType == .selected ? .on : .off
        watchedFolderTypeButtonAll?.state = watchedFolderType == .all  ? .on : .off
        watchedFolderTypeButtonSpecific?.state = watchedFolderType == .specific ? .on : .off
    }
}
